File: Runtime/Plugins/iOS/NuxiePurchaseDelegateBridge.swift
import Foundation

#if canImport(Nuxie)
import Nuxie

final class NuxiePurchaseDelegateBridge: NuxiePurchaseDelegate, @unchecked Sendable {
  private let emit: (String, [String: Any]) -> Void
  private let timeoutSeconds: TimeInterval
  private let lock = NSLock()
  private var purchaseContinuations: [String: CheckedContinuation<PurchaseResult, Never>] = [:]
  private var restoreContinuations: [String: CheckedContinuation<RestoreResult, Never>] = [:]

  init(
    timeoutSeconds: TimeInterval = 60,
    emit: @escaping (String, [String: Any]) -> Void
  ) {
    self.timeoutSeconds = timeoutSeconds
    self.emit = emit
  }

  func purchase(product: StoreProduct) async -> PurchaseResult {
    let requestId = UUID().uuidString
    let payload: [String: Any] = [
      "requestId": requestId,
      "product": [
        "platform": "ios", "productId": product.productId, "storeProductId": product.storeProductId,
        "placementId": product.placementId, "displayName": product.name, "displayPrice": product.price,
        "description": product.description, "productType": product.productType.rawValue,
        "period": nuxieNullable(product.period?.rawValue), "periodCount": nuxieNullable(product.periodCount),
        "eligibilityJws": nuxieNullable(product.introductoryOfferEligibilityJWS), "billingPlan": product.billingPlan.rawValue,
        "introductoryTerms": nuxieNullable(product.introductoryTerms.map { terms in [
          "price": terms.price, "period": terms.period.rawValue, "periodCount": terms.periodCount,
          "cycles": terms.cycles, "paymentMode": terms.paymentMode.rawValue, "displayDuration": terms.trialPeriodText,
        ] as [String: Any] }),
      ] as [String: Any],
    ]

    return await withCheckedContinuation { continuation in
      lock.withLock {
        purchaseContinuations[requestId] = continuation
      }
      emit("purchase", payload)
      schedulePurchaseTimeout(requestId: requestId)
    }
  }

  func restorePurchases() async -> RestoreResult {
    let requestId = UUID().uuidString
    let payload: [String: Any] = [
      "requestId": requestId,
      "platform": "ios",
      "timestamp_ms": Int(Date().timeIntervalSince1970 * 1_000),
    ]

    return await withCheckedContinuation { continuation in
      lock.withLock {
        restoreContinuations[requestId] = continuation
      }
      emit("restore", payload)
      scheduleRestoreTimeout(requestId: requestId)
    }
  }

  func completePurchase(requestId: String, payload: [String: Any]) {
    let continuation = lock.withLock {
      purchaseContinuations.removeValue(forKey: requestId)
    }
    continuation?.resume(returning: purchaseResult(from: payload))
  }

  func completeRestore(requestId: String, payload: [String: Any]) {
    let continuation = lock.withLock {
      restoreContinuations.removeValue(forKey: requestId)
    }
    continuation?.resume(returning: restoreResult(from: payload))
  }

  func cancelPending() {
    let (purchases, restores) = lock.withLock {
      let purchases = Array(purchaseContinuations.values)
      let restores = Array(restoreContinuations.values)
      purchaseContinuations.removeAll()
      restoreContinuations.removeAll()
      return (purchases, restores)
    }
    purchases.forEach { $0.resume(returning: .failed(bridgeError("sdk_shutdown"))) }
    restores.forEach { $0.resume(returning: .failed(bridgeError("sdk_shutdown"))) }
  }

  private func schedulePurchaseTimeout(requestId: String) {
    Task { [weak self] in
      guard let self else { return }
      try? await Task.sleep(nanoseconds: UInt64(self.timeoutSeconds * 1_000_000_000))
      let continuation = self.lock.withLock {
        self.purchaseContinuations.removeValue(forKey: requestId)
      }
      continuation?.resume(returning: .failed(self.bridgeError("purchase_timeout")))
    }
  }

  private func scheduleRestoreTimeout(requestId: String) {
    Task { [weak self] in
      guard let self else { return }
      try? await Task.sleep(nanoseconds: UInt64(self.timeoutSeconds * 1_000_000_000))
      let continuation = self.lock.withLock {
        self.restoreContinuations.removeValue(forKey: requestId)
      }
      continuation?.resume(returning: .failed(self.bridgeError("restore_timeout")))
    }
  }

  private func purchaseResult(from payload: [String: Any]) -> PurchaseResult {
    switch (payload["type"] as? String) {
    case "purchased": .purchased
    case "cancelled": .cancelled
    case "pending": .pending
    default: .failed(bridgeError((payload["message"] as? String) ?? "purchase_failed"))
    }
  }

  private func restoreResult(from payload: [String: Any]) -> RestoreResult {
    switch (payload["type"] as? String) {
    case "restored": .restored
    case "noPurchases": .noPurchases
    default: .failed(bridgeError((payload["message"] as? String) ?? "restore_failed"))
    }
  }

  private func bridgeError(_ message: String) -> Error {
    NSError(
      domain: "io.nuxie.unity",
      code: 1,
      userInfo: [NSLocalizedDescriptionKey: message]
    )
  }
}
#endif

File: Native~/Android/src/main/kotlin/ai/nuxie/unity/NuxiePurchaseDelegateBridge.kt
package ai.nuxie.unity

import ai.nuxie.sdk.billing.NuxiePurchaseDelegate
import ai.nuxie.sdk.billing.PurchaseResult
import ai.nuxie.sdk.billing.RestoreResult
import ai.nuxie.sdk.billing.StoreProduct
import java.util.UUID
import java.util.concurrent.ConcurrentHashMap
import kotlinx.coroutines.CompletableDeferred
import kotlinx.coroutines.withTimeoutOrNull

class NuxiePurchaseDelegateBridge(
  private val emitEvent: (eventName: String, payload: Map<String, Any?>) -> Unit,
  private val timeoutMs: Long = 60_000,
) : NuxiePurchaseDelegate {
  private val purchaseRequests = ConcurrentHashMap<String, CompletableDeferred<PurchaseResult>>()
  private val restoreRequests = ConcurrentHashMap<String, CompletableDeferred<RestoreResult>>()

  override suspend fun purchase(product: StoreProduct): PurchaseResult {
    val requestId = UUID.randomUUID().toString()
    val deferred = CompletableDeferred<PurchaseResult>()
    purchaseRequests[requestId] = deferred

    emitEvent(
      "purchase",
      mapOf(
        "requestId" to requestId,
        "product" to mapOf(
          "platform" to "android", "productId" to product.productId, "storeProductId" to product.storeProductId,
          "basePlanId" to product.basePlanId, "purchaseOptionId" to product.purchaseOptionId,
          "offerId" to product.offerId, "placementId" to product.placementId,
          "displayName" to product.rawProduct?.name, "description" to product.rawProduct?.description,
          "productType" to product.rawProduct?.productType,
          "displayPrice" to selectedPrice(product),
          "pricingPhases" to product.rawProduct?.subscriptionOfferDetails
            ?.firstOrNull { it.basePlanId == product.basePlanId && it.offerId == product.offerId }
            ?.pricingPhases?.pricingPhaseList?.map { phase -> mapOf(
              "displayPrice" to phase.formattedPrice, "billingPeriod" to phase.billingPeriod,
              "billingCycleCount" to phase.billingCycleCount, "recurrenceMode" to phase.recurrenceMode,
            ) },
        ),
        "timestamp_ms" to System.currentTimeMillis(),
      ),
    )

    return try {
      withTimeoutOrNull(timeoutMs) { deferred.await() }
        ?: PurchaseResult.Failed(bridgeError("purchase_timeout"))
    } finally {
      purchaseRequests.remove(requestId)
    }
  }

  override suspend fun restorePurchases(): RestoreResult {
    val requestId = UUID.randomUUID().toString()
    val deferred = CompletableDeferred<RestoreResult>()
    restoreRequests[requestId] = deferred

    emitEvent(
      "restore",
      mapOf(
        "requestId" to requestId,
        "platform" to "android",
        "timestamp_ms" to System.currentTimeMillis(),
      ),
    )

    return try {
      withTimeoutOrNull(timeoutMs) { deferred.await() }
        ?: RestoreResult.Failed(bridgeError("restore_timeout"))
    } finally {
      restoreRequests.remove(requestId)
    }
  }

  fun completePurchase(requestId: String, payload: Map<String, Any?>) {
    purchaseRequests.remove(requestId)?.complete(purchaseResult(payload))
  }

  fun completeRestore(requestId: String, payload: Map<String, Any?>) {
    restoreRequests.remove(requestId)?.complete(restoreResult(payload))
  }

  fun cancelPending(reason: String) {
    purchaseRequests.values.forEach { request ->
      request.complete(PurchaseResult.Failed(bridgeError(reason)))
    }
    restoreRequests.values.forEach { request ->
      request.complete(RestoreResult.Failed(bridgeError(reason)))
    }
    purchaseRequests.clear()
    restoreRequests.clear()
  }

  private fun purchaseResult(payload: Map<String, Any?>): PurchaseResult =
    when ((payload["type"] as? String)) {
      "purchased" -> PurchaseResult.Purchased
      "cancelled" -> PurchaseResult.Cancelled
      "pending" -> PurchaseResult.Pending
      else -> PurchaseResult.Failed(
        bridgeError((payload["message"] as? String) ?: "purchase_failed"),
      )
    }

  private fun restoreResult(payload: Map<String, Any?>): RestoreResult =
    when ((payload["type"] as? String)) {
      "restored" -> RestoreResult.Restored
      "noPurchases" -> RestoreResult.NoPurchases
      else -> RestoreResult.Failed(
        bridgeError((payload["message"] as? String) ?: "restore_failed"),
      )
    }

  private fun selectedPrice(product: StoreProduct): String? {
    val details = product.rawProduct ?: return null
    return if (details.productType == "subs") details.subscriptionOfferDetails
      ?.firstOrNull { it.basePlanId == product.basePlanId && it.offerId == product.offerId }
      ?.pricingPhases?.pricingPhaseList?.lastOrNull()?.formattedPrice
    else details.oneTimePurchaseOfferDetailsList
      ?.firstOrNull { it.purchaseOptionId == product.purchaseOptionId && it.offerId == product.offerId }
      ?.formattedPrice
  }

  private fun bridgeError(message: String): Throwable = IllegalStateException(message)
}
